src/Operation/FindOneAndReplace.php

```diff
@@ -177,4 +170,34 @@ public function getCommandDocument()
     {
         return $this->findAndModify->getCommandDocument();
     }
+
+    /**
+     * @param array|object $replacement
+     * @return array|object
+     */
+    private function validateReplacement($replacement, ?DocumentCodec $codec)
+    {
+        if (isset($codec)) {
+            $replacement = $codec->encode($replacement);
+        }
+
+        if (! is_document($replacement)) {
+            throw InvalidArgumentException::expectedDocumentType('$replacement', $replacement);
+        }
+
+        // Treat empty arrays as replacement documents for BC
+        if ($replacement === []) {
+            $replacement = (object) $replacement;
+        }
+
+        if (is_first_key_operator($replacement)) {
+            throw new InvalidArgumentException('First key in $replacement is an update operator');
```

Comment:
Related to my other question about creating common utility functions to validate the replacement and insert documents, I probably overlooked that we want to provide contextual exception messages here. So in that case I won't object to keeping some of this duplicated logic as-is.